src/coreclr/jit/promotion.cpp

```diff
@@ -743,31 +783,551 @@ class ReplaceVisitor : public GenTreeVisitor<ReplaceVisitor>
+                    dstFirstRep++;
+                }
+
+                if (dstEndRep > dstFirstRep)
+                {
+                    Replacement* dstLastRep = dstEndRep - 1;
+                    if (dstLastRep->Offset + genTypeSize(dstLastRep->AccessType) > dstLclOffs + dstLclSize)
+                    {
+                        JITDUMP("*** Block operation partially overlaps with %s. Write and read-backs are necessary.\n",
+                                dstLastRep->Name);
+                        result.AddStatement(CreateWriteBack(dstLcl->GetLclNum(), *dstLastRep));
+
+                        dstLastRep->NeedsWriteBack = false;
+                        dstLastRep->NeedsReadBack  = true;
+                        dstEndRep--;
+                    }
+                }
+
+                if (src->IsConstInitVal())
+                {
+                    GenTree* cns = src->OperIsInitVal() ? src->gtGetOp1() : src;
+                    InitFieldByField(dstFirstRep, dstEndRep, static_cast<unsigned char>(cns->AsIntCon()->IconValue()),
+                                     &result);
+                }
+                else
+                {
+                    CopyIntoFields(dstFirstRep, dstEndRep, dstLcl, src, &result);
+                }
+
+                // At this point all replacements that have Handled = true contain their correct value.
+                // Check if these cover the entire block operation.
+                unsigned prevEnd = dstLclOffs;
+                bool     covered = true;
+
+                for (Replacement* rep = dstFirstRep; rep < dstEndRep; rep++)
+                {
+                    if (!rep->Handled)
+                    {
+                        covered = false;
+                        break;
+                    }
+
+                    assert(rep->Offset >= prevEnd);
+                    if (rep->Offset != prevEnd)
+                    {
+                        // Uncovered hole from [lastEnd..rep->Offset).
+                        // TODO-CQ: In many cases it's more efficient to "plug" the holes. However,
+                        // it is made more complicated by the fact that the holes can contain GC pointers in them and
+                        // we cannot (yet) represent custom class layouts with GC pointers in them.
+                        // TODO-CQ: Many of these cases are just padding. We should handle structs with insignificant
+                        // padding here.
+                        covered = false;
+                        break;
+                    }
+
+                    prevEnd = rep->Offset + genTypeSize(rep->AccessType);
+                }
+
+                covered &= prevEnd == dstLclOffs + dstLclSize;
+
+                if (!covered)
+                {
+                    JITDUMP("Struct operation is not fully covered by replaced fields. Keeping struct operation.\n");
+                    result.AddStatement(asg);
```

Comment:
I think I am confused here -- it seems like `result` already accumulated IR for the handled cases and now we're appending a full copy, but some of the copied bits may be stale if we didn't write them back to the source -- but we know not to read from the corresponding bits in the destination?

If so maybe expand on the comment below to say "For handled replacements, we must not read back..."

Reply:
The copied bits won't be stale here since this PR always (conservatively) writes the source replacements back to its struct local to begin with, but yes, for the replacements that were handled we won't read them back from the struct local due to the loop below, so it wouldn't matter even if they were.

Once we get to handling sources this "covering" copy will need to happen before the copying of the source replacements since otherwise we would overwrite the fresh bits (from the source's replacements) with stale bits (from the source's struct local).